flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/StateChangeLoggerImpl.java

```diff
@@ -0,0 +1,148 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.state.changelog;
+
+import org.apache.flink.api.common.typeutils.TypeSerializer;
+import org.apache.flink.core.memory.DataOutputViewStreamWrapper;
+import org.apache.flink.runtime.state.changelog.StateChangelogWriter;
+import org.apache.flink.runtime.state.heap.InternalReadOnlyKeyContext;
+import org.apache.flink.util.ExceptionUtils;
+import org.apache.flink.util.function.ThrowingConsumer;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.util.Collection;
+
+import static org.apache.flink.state.changelog.StateChangeLoggerImpl.StateChangeOperation.ADD;
+import static org.apache.flink.state.changelog.StateChangeLoggerImpl.StateChangeOperation.CHANGE_ELEMENT;
+import static org.apache.flink.state.changelog.StateChangeLoggerImpl.StateChangeOperation.CLEAR;
+import static org.apache.flink.state.changelog.StateChangeLoggerImpl.StateChangeOperation.MERGE_NS;
+import static org.apache.flink.state.changelog.StateChangeLoggerImpl.StateChangeOperation.REMOVE_ELEMENT;
+import static org.apache.flink.state.changelog.StateChangeLoggerImpl.StateChangeOperation.SET;
+
+class StateChangeLoggerImpl<Key, State, Ns> implements StateChangeLogger<State, Ns> {
+
+    private final StateChangelogWriter<?> stateChangelogWriter;
+    private final InternalReadOnlyKeyContext<Key> keyContext;
+    private final TypeSerializer<Key> keySerializer;
+    private final TypeSerializer<Ns> namespaceSerializer;
+    private final TypeSerializer<State> valueSerializer;
+
+    public StateChangeLoggerImpl(
+            TypeSerializer<Key> keySerializer,
+            TypeSerializer<Ns> namespaceSerializer,
+            TypeSerializer<State> valueSerializer,
+            InternalReadOnlyKeyContext<Key> keyContext,
+            StateChangelogWriter<?> stateChangelogWriter) {
+        this.stateChangelogWriter = stateChangelogWriter;
+        this.keyContext = keyContext;
+        this.valueSerializer = valueSerializer;
+        this.keySerializer = keySerializer;
+        this.namespaceSerializer = namespaceSerializer;
+    }
+
+    @Override
+    public void stateUpdated(State newState, Ns ns) throws IOException {
+        if (newState == null) {
+            stateCleared(ns);
+        } else {
+            log(serialize(SET, ns, out -> valueSerializer.serialize(newState, out)));
+        }
+    }
+
+    @Override
+    public void stateAdded(State addedState, Ns ns) throws IOException {
+        log(serialize(ADD, ns, out -> valueSerializer.serialize(addedState, out)));
+    }
+
+    @Override
+    public void stateCleared(Ns ns) {
+        try {
+            log(serialize(CLEAR, ns, out -> {}));
+        } catch (IOException e) {
+            ExceptionUtils.rethrow(e);
+        }
+    }
```

Comment:
This is because `clear()` method on State classes doesn't declare any checked exceptions. Declaring it here would increase this boilerplate code (though I agree that API here is non-uniform in that respect).